### ISSUE TITLE
Multiple browser refreshes to establish rTorrent connection

### ISSUE THREAD

**Issue by Do-while-Bacon** [jfurrow/flood#713]
## Summary
Several browser refreshes are needed to establish a connection to rTorrent using UNIX socket.

EDIT: in Firefox

## Expected Behavior
Flood should load on the first try.

## Current Behavior
Every time I visit Flood, I have to CTRL+F5 between 4-8 times to get the torrent status page to load. It hangs at this stage:

![Image](https://i.imgur.com/Z2ZDSzT.jpg)

If I leave it there long enough, it will eventually time out with "Cannot connect to rTorrent." Rapidly refreshing the page will eventually get it to connect. When it does connect, there is no delay at all.

## Steps to Reproduce 
1. Configure rTorrent to use a UNIX socket
2. Configure Flood to point to that socket
3. NGINX reverse proxy
4. Access with Firefox

## Your Environment
CentOS 7.5.1804
Node.js version 10.10.0
Flood commit 8867c7d3df16fdccbb4d882a0ea042d49f0b64c6
rTorrent version 0.9.6-2 (epel release)
NGINX (reverse proxy) version 1.12.2-2 (epel release)
Firefox 62

Flood config.js:
```
const CONFIG = {
  baseURI: '/flood',
  dbCleanInterval: 1000 * 60 * 60,
  dbPath: './server/db/',
  floodServerHost: '0.0.0.0',
  floodServerPort: 3000,
  floodServerProxy: 'http://127.0.0.1:3000',
  maxHistoryStates: 30,
  torrentClientPollInterval: 1000 * 2,
  secret: '{REDACTED}',
  ssl: false,
  sslKey: '/absolute/path/to/key/',
  sslCert: '/absolute/path/to/certificate/'
};
```

rTorrent config (relevant lines):
```
network.scgi.open_local = /run/rtorrent/.rtorrent.sock
schedule2 = chmod_scgi_socket, 0, 0, "execute2=chmod,\"g+w,o=\",/run/rtorrent/.rtorrent.sock"
```

NGINX config (relevant lines):
```
        location /flood/ {
            proxy_pass http://127.0.0.1:3000/;
            allow 192.168.1.0/24;
            allow 192.168.2.0/24;
            deny all;
        }
```

I have confirmed that the "flood" daemon user has permission to read and write the rtorrent.sock (of course it would never work if it didn't).

Both rtorrent and Flood are configured as systemd services.

Thanks for any advice you can offer.

EDIT: After further testing, it appears that this issue only happens on Firefox. I have been unable to reproduce it on Chrome.

**Comment by bbashy** [comment]
I also get this sometimes. And then there's times when the UI is reset (column widths) and I have to refresh to get my settings.

Maybe something is loading too fast and is waiting on something that has already happened (normal JS stuff).

**Comment by noraj** [comment]
@Do-while-Bacon Did you tried to restart Firefox in debug mode with all plugins and extensions disabled so that you can see if teh problem is not coming from one of them ?

**Comment by Do-while-Bacon** [comment]
@noraj Thanks for the good suggestion.

It's strange, actually, The problem is gone when I run Firefox in Safe Mode, but, if I manually disable all of my extensions and run in normal mode, the problem is still there. 

Do you know what else Safe Mode does besides disabling all extensions? I also tried disabling hardware acceleration, but, that was not it.

I only run a few extensions and they are very popular -- uBlock Origin, LastPass...etc.

**Comment by noraj** [comment]
`uBlock Origin` blocks a lot more than just ads and caused me some unexpected problems for some websites. But I'm using it and have no problem with Flood so it can't be it.

Yeah safe mode is disabling extension and hardware acceleration but also reset some parameters. Have you changed some params in `about:config`? Or maybe you are using a custom theme?

**Comment by Do-while-Bacon** [comment]
@noraj My Firefox is mostly stock -- no special configs or themes, but, just to be extra safe I did a complete wipe of my Firefox profile and restored it to default state. No addons, no settings, just 100% fresh install. However, I am still having the problem with Flood. Now I am quite confident this is not a problem with my Firefox.

Also, I did some more testing with Safe Mode. The problem actually does appear in Safe Mode too, but far less frequently. Maybe about 1 in 10 refresh will hang and fail to connect. I don't know why it happens less often in Safe Mode, even with 100% fresh browser.

**Comment by jfurrow** [comment]
It sounds like a race condition somewhere... will investigate. I've not been able to reproduce, for what it's worth.

**Comment by bbashy** [comment]
It does it more in Firefox for me (plain, never changed any settings or added any extensions).

It's always the top one that doesn't complete (visually).

![](https://d.code.im/QhJOaj15uD3rSLQltzNJVTOEkxLSu1ructELU1s9PR.jpg)

```
The connection to https://code.im/api/activity-stream?historySnapshot=fiveMin was interrupted while the page was loading. main.be5d4735.js:1:279423
Navigated to https://code.im/overview
The connection to https://code.im/api/activity-stream?historySnapshot=fiveMin was interrupted while the page was loading. FloodActions.js:273:34
Navigated to https://code.im/overview
The connection to https://code.im/api/activity-stream?historySnapshot=fiveMin was interrupted while the page was loading.
```

**Comment by noraj** [comment]
@Do-while-Bacon 
I commented `floodServerProxy` because it is only for dev and on the past it made bug my app.

```
// Used for development. See the "Local Development" section of README.md
  // for detail.
  /*floodServerProxy: 'https://127.0.0.1',*/
```

Also I suggest you to put `floodServerHost: '127.0.0.1',` you don't need to expose all interfaces, only localhost is needed for the reverse proxy. This will be more secure.

For rtorrent connection I'm using a TCP socket `scgi_port = 127.0.0.1:5000`, it working. I know a UNIX socket is more secure, so if you make it work with a UNIX socket I'm interested to know the working conf.

In your nginx config try to change your `proxy_pass` removing the trailing slash. I also have a rewrite rule `rewrite ^/(.*) /$1 break;` but I don't really if it is needed anymore.

**Comment by Do-while-Bacon** [comment]
It's very kind of you guys to help me with this. I appreciate it.

@jfurrow It does seem like a race condition of some kind. I have not been able to figure out what causes me to win or lose the race. I tried Flood from my corporate laptop (MacBook w/ Firefox 60 ESR) and it was much more successful -- worked on about 75% of page reloads. My Windows desktop is about 15% success rate. Please let me know if there is any debug information I can provide or tests you'd like me to try. I'm sorry that I don't have much of any developer experience to contribute.

@noraj 
> I commented `floodServerProxy` because it is only for dev and on the past it made bug my app.
Thanks. I've done the same now. I wasn't really sure if I needed that or not.

> Also I suggest you to put `floodServerHost: '127.0.0.1',` you don't need to expose all interfaces, only localhost is needed for the reverse proxy. This will be more secure.
I firewall all ports except 80/443 to this server, but, you are right that it does not make sense to listen for outside connections. Fixed.

I did an npm install/build after those two changes.

> I know a UNIX socket is more secure, so if you make it work with a UNIX socket I'm interested to know the working conf.
For sure it works if you are a Chrome user. :) Though, actually, I haven't tried a TCP socket with Flood, so I don't know if the problem is only with UNIX socket. I should test that -- I'm just very paranoid about security. Can I change that setting without adding a new user?

> In your nginx config try to change your `proxy_pass` removing the trailing slash. I also have a rewrite rule `rewrite ^/(.*) /$1 break;` but I don't really if it is needed anymore
It did not load at all without the trailing slash. :(

**Comment by williamkray** [comment]
i get similar issues with firefox, to the point where the flood interface is loading just fine but says "Cannot connect to rtorrent"

works just fine in chromium.

running rtorrent in a docker container with xmlrpc on a network port, with flood in a separate container on the same docker network communicating to the docker container via the port. flood is behind an nginx proxy in (you guessed it) a docker container.

**Comment by williamkray** [comment]
i just tested in a fresh Firefox profile with no addons, and it worked a little better than my normal profile... but even refreshing once got me to hang at this point.
![image](https://user-images.githubusercontent.com/8409433/46046848-317cac80-c0e8-11e8-85de-564ac2a5cf72.png)

**Comment by jfurrow** [comment]
I've observed the settings failure myself but haven't figured out how to reproduce. Thanks all for the good info in this thread, I'll continue to investigate.

**Comment by Scandiskfr** [comment]
Hi all,
Same pb for me, i juste update this morning and i'm cannot connect with Firefox or Edge.
Chrome is working well.
On edge i have only blue page and nothing more.
On Firefox, i can enter my auth but stay at the same screen shot like previus post (williamkray).

I not using nginx or proxy juste Apache/2.4.18 (Ubuntu)
Thx for your help.

**Comment by Scandiskfr** [comment]
Non reverse proxy, sorry it's a direct Access.

**Comment by noraj** [comment]
@Scandiskfr So what do you mean by "I'm just using Apache/2.4.18" ?

**Comment by Scandiskfr** [comment]
Another detail, alwayse in Firefox first connexion working ! but if i refresh the page or reconnect two minuts later it's stuck ! but the Firefox tab show me the speed. If i wait a very long time in this stage, Flood loading but display "cannot connect to rtorrent".
In IE or Edge i juste have blue page nothing more.
All working good with Chrome.
I hop this detail help you.

**Comment by bbashy** [comment]
@Scandiskfr Node will use its own server — saying you’re using Apache, means it’s a proxy to node.

**Comment by mirtouf** [comment]
https://github.com/jfurrow/flood/issues/713#issuecomment-423489126

Same here, this issue does not happen with Chromium. I will investigate.

**Comment by jfurrow** [comment]
I still can't reproduce, but try this PR: https://github.com/jfurrow/flood/pull/721 (make sure to run `npm run build` to build the new app bundle).

I think some people in this thread might just have their XMLRPC connection to rtorrent misconfigured, but the above PR might address a race condition with the settings request failing too early... maybe?

**Comment by mirtouf** [comment]
I checked out the branch bug/fix-settings-fetch-failure, rebuild the assets, reload the application and I still have this error with firefox :
``
The connection to https://bla/api/activity-stream?historySnapshot=fiveMin was interrupted while the page was loading. main.8d1db5ef.js:1:279439
``

**Comment by jfurrow** [comment]
@mirtouf That sounds like a separate issue to me, honestly.

I just pushed another commit to [the PR](https://github.com/jfurrow/flood/pull/721) and I'm fairly confident that's the fix. I am able to reliably reproduce in `master` and that PR fixes it for me.

**Comment by mirtouf** [comment]
It tastes like #391 indeed.

**Comment by mirtouf** [comment]
It seems that text/event-stream is considered as a download for Firefox whereas Chromium is displaying the output correctly.
I will seek if this case can be mitigated through nginx configuration.

**Comment by Do-while-Bacon** [comment]
#721 did not fix the Firefox issues for me. :(

**Comment by jfurrow** [comment]
@Do-while-Bacon Can you send a screenshot of the network requests?

You can do this by opening the developer tools, then click the "Network" tab, then hard-reload the Flood UI. It should look something like this:
<img width="558" alt="screen shot 2018-09-30 at 10 47 39 am" src="https://user-images.githubusercontent.com/1134755/46260713-55e8d800-c49e-11e8-8b0e-98a1dc214235.png">

**Comment by Do-while-Bacon** [comment]
**Failed page load:**
![fail](https://user-images.githubusercontent.com/9923207/46261464-fdc3cd00-c4c1-11e8-979c-3e310d52ec3f.jpg)

**Successful page load:**
![success](https://user-images.githubusercontent.com/9923207/46261473-04524480-c4c2-11e8-932d-47ccf46dc389.jpg)

**Comment by mirtouf** [comment]
To evacuate the reverse proxy eventual problems, I set-up a fresh installation with the up-to-date git and I got a 500 error while trying to save the parameters.
I ran
``
npm run start:development:server 
> flood@1.0.0 start:development:server /srv/flood
> NODE_ENV=development nodemon server/bin/start.js
[nodemon] 1.17.4
[nodemon] to restart at any time, enter `rs`
[nodemon] watching: *.*
[nodemon] starting `node server/bin/start.js`
Flood server starting on http://0.0.0.0:3333.
POST /api/client/connection-test 200 103.569 ms - 20
PATCH /auth/users/mirtouf 500 1138.513 ms - -
PATCH /auth/users/mirtouf 500 86.895 ms - -
POST /api/client/connection-test 200 15.333 ms - 20
POST /api/client/connection-test 200 18.547 ms - 20
PATCH /auth/users/mirtouf 500 76.213 ms - -
``

<!DOCTYPE html><html><head><title>Flood Error</title></head><body><h1>Users.patchUser is not a function</h1><h2></h2><pre>TypeError: Users.patchUser is not a function
at router.patch (/srv/flood/server/routes/auth.js:158:9)
at Layer.handle [as handle_request] (/srv/flood/node_modules/express/lib/router/layer.js:95:5)
at next (/srv/flood/node_modules/express/lib/router/route.js:137:13)
at Route.dispatch (/srv/flood/node_modules/express/lib/router/route.js:112:3)
at Layer.handle [as handle_request] (/srv/flood/node_modules/express/lib/router/layer.js:95:5)
at /srv/flood/node_modules/express/lib/router/index.js:281:22
at param (/srv/flood/node_modules/express/lib/router/index.js:354:14)
at param (/srv/flood/node_modules/express/lib/router/index.js:365:14)
at Function.process_params (/srv/flood/node_modules/express/lib/router/index.js:410:3)
at next (/srv/flood/node_modules/express/lib/router/index.js:275:10)</pre></body></html>

I confirm I got the same behaviour described by Do-while-Bacon, there is no difference between a successful loading and a failed one when proxying flood.

**Comment by mirtouf** [comment]
OK, after some investigation I found out that if I log out and then log in, connection to rtorrent seems to be 100% successful.

**Comment by Scandiskfr** [comment]
Hi all,
After ugrading to last version all is working good on firefox ! but stay "blue" page on Edge, no login (it's not a problem for me, i don't use Edge LOL)
Thx for you job ;)

**Comment by gerwalk** [comment]
I've got the same issue - works fine in Chrome, rarely works in Firefox. Reverse-proxy from nginx. On the latest from master so PR 721 didn't solve it.

@jfurrow 
> **Edit** I was able to reproduce on a remote server with Firefox (still not able to reproduce locally) 

Wonder if this a key detail. I am also trying to access remotely - rtorrent/nginx/flood are on my headless web server box which I SSH to from my main desktop or other devices.

**Comment by EndlessDex** [comment]
@mirtouf suggestion to logout and then back in worked for me while experiencing this problem.

**Comment by jfurrow** [comment]
@gerwalk I was able to repro and fix the remote server issue I experienced. I am still unable to repro the other issues referred to in this thread.

**Comment by bbashy** [comment]
I believe the original problem for this issue has been resolved.
